Backfill_Correction/delphiBackfillCorrection/R/preprocessing.R

```diff
@@ -0,0 +1,222 @@
+#' @template df-template
+#' @template value_col-template
+#' @template refd_col-template
+#' @template lag_col-template
+#' 
+#' @importFrom tidyr fill pivot_wider pivot_longer
+#' @importFrom dplyr %>% everything select
+#' 
+#' @export
+fill_missing_updates <- function(df, value_col, refd_col, lag_col) {
+  pivot_df <- df[order(df[[lag_col]], decreasing=FALSE), ] %>%
+    pivot_wider(id_cols=lag_col, names_from=refd_col, values_from=value_col)
+  
+  if (any(diff(pivot_df[[lag_col]]) != 1)) {
+    stop("Risk exists in forward filling")
+  }
+  pivot_df <- pivot_df %>% fill(everything(), .direction="down")
+  
+  # Fill NAs with 0s
+  pivot_df[is.na(pivot_df)] <- 0
+  
+  backfill_df <- pivot_df %>%
+    pivot_longer(-lag_col, values_to="value_raw", names_to=refd_col)
+  backfill_df[[refd_col]] = as.Date(backfill_df[[refd_col]])
+  
+  return (as.data.frame(backfill_df))
+}
+
+#' Calculate 7 day moving average for each issue date
+#' The 7dav for date D reported on issue date D_i is the average from D-7 to D-1
+#' @param pivot_df Data Frame where the columns are issue dates and the rows are 
+#'    reference dates
+#' @template refd_col-template
+#' 
+#' @importFrom zoo rollmeanr
+#' 
+#' @export
+get_7dav <- function(pivot_df, refd_col) {
+  for (col in colnames(pivot_df)) {
+    if (col == refd_col) next
+    pivot_df[, col] <- rollmeanr(pivot_df[, col], 7, align="right", fill=NA)
+  }
+  backfill_df <- pivot_df %>%
+    pivot_longer(-refd_col, values_to="value_raw", names_to="issue_date")
+  backfill_df[[refd_col]] = as.Date(backfill_df[[refd_col]])
+  backfill_df[["issue_date"]] = as.Date(backfill_df[["issue_date"]])
+  return (as.data.frame(backfill_df))
+}
+
+#' Used for data shifting in terms of reference date
+#' 
+#' @template df-template
+#' @param n_day number of days to be shifted
+#' @template refd_col-template
+#' 
+#' @export
+add_shift <- function(df, n_day, refd_col) {
+  df[, refd_col] <- as.Date(df[, refd_col]) + n_day
+  return (df)
+}
+
+#' Add one hot encoding for day of a week info in terms of reference
+#' and issue date
+#' 
+#' @template df-template
+#' @param wd vector of days of a week
+#' @template time_col-template
+#' @param suffix suffix added to indicate which kind of date is used
+#' 
+#' @export
+add_dayofweek <- function(df, time_col, suffix, wd = WEEKDAYS_ABBR) {
+  dayofweek <- as.numeric(format(df[[time_col]], format="%u"))
+  for (i in 1:6) {
+    df[, paste0(wd[i], suffix)] <- as.numeric(dayofweek == i)
+  }
+  if (suffix == "_ref") {
+    df[, paste0("Sun", suffix)] <- as.numeric(dayofweek == 7)
+  }
+  return (df)
+}
+
+#' Get week of a month info according to a date
+#'
+#' All the dates on or before the ith Sunday but after the (i-1)th Sunday
+#' is considered to be the ith week. Notice that the dates in the 5th week
+#' this month are actually in the same week with the dates in the 1st week
+#' next month and those dates are sparse. Thus, we assign the dates in the 
+#' 5th week to the 1st week.
```

Comment:
So are October 2022 weeks numbered like this:

```
𝙾𝙲𝚃𝙾𝙱𝙴𝚁 – 𝟸𝟶𝟸𝟸
𝚂𝚞|𝙼𝚘|𝚃𝚞|𝚆𝚎|𝚃𝚑|𝙵𝚛|𝚂𝚊
------------------
𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝟶𝟷 week 1
𝟶𝟸|𝟶𝟹|𝟶𝟺|𝟶𝟻|𝟶𝟼|𝟶𝟽|𝟶𝟾 week 2
𝟶𝟿|𝟷𝟶|𝟷𝟷|𝟷𝟸|𝟷𝟹|𝟷𝟺|𝟷𝟻 week 3
𝟷𝟼|𝟷𝟽|𝟷𝟾|𝟷𝟿|𝟸𝟶|𝟸𝟷|𝟸𝟸 week 4
𝟸𝟹|𝟸𝟺|𝟸𝟻|𝟸𝟼|𝟸𝟽|𝟸𝟾|𝟸𝟿 week 1 November
𝟹𝟶|𝟹𝟷|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡 week 2 November
```

?

Reply:
No, it should be
```
𝙾𝙲𝚃𝙾𝙱𝙴𝚁 – 𝟸𝟶𝟸𝟸
𝚂𝚞|𝙼𝚘|𝚃𝚞|𝚆𝚎|𝚃𝚑|𝙵𝚛|𝚂𝚊
------------------
𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝟶𝟷 week 1
𝟶𝟸|𝟶𝟹|𝟶𝟺|𝟶𝟻|𝟶𝟼|𝟶𝟽|𝟶𝟾 week 1
𝟶𝟿|𝟷𝟶|𝟷𝟷|𝟷𝟸|𝟷𝟹|𝟷𝟺|𝟷𝟻 week 2
𝟷𝟼|𝟷𝟽|𝟷𝟾|𝟷𝟿|𝟸𝟶|𝟸𝟷|𝟸𝟸 week 3
𝟸𝟹|𝟸𝟺|𝟸𝟻|𝟸𝟼|𝟸𝟽|𝟸𝟾|𝟸𝟿 week 4
𝟹𝟶|𝟹𝟷|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡|𝚡𝚡 week 1 November
```
Sorry this function is not up-to-date, updated just now.